lib/route/cls/flower.c

```diff
@@ -0,0 +1,779 @@
+/*
+ * lib/route/cls/flower.c		Flow based traffic control filter
+ *
+ *	This library is free software; you can redistribute it and/or
+ *	modify it under the terms of the GNU Lesser General Public
+ *	License as published by the Free Software Foundation version 2.1
+ *	of the License.
+ *
+ * Copyright (c) 2018 Volodymyr Bendiuga <volodymyr.bendiuga@gmail.com>
+ */
+
+#include <netlink-private/netlink.h>
+#include <netlink-private/tc.h>
+#include <netlink/netlink.h>
+#include <netlink/attr.h>
+#include <netlink/utils.h>
+#include <netlink-private/route/tc-api.h>
+#include <netlink/route/classifier.h>
+#include <netlink/route/action.h>
+#include <netlink/route/cls/flower.h>
+
+
+/** @cond SKIP */
+#define FLOWER_ATTR_FLAGS           1 << 0x0
+#define FLOWER_ATTR_ACTION          1 << 0x1
+#define FLOWER_ATTR_POLICE          1 << 0x2
+#define FLOWER_ATTR_VLAN_ID         1 << 0x3
+#define FLOWER_ATTR_VLAN_PRIO       1 << 0x4
+#define FLOWER_ATTR_VLAN_ETH_TYPE   1 << 0x5
+#define FLOWER_ATTR_DST_MAC         1 << 0x6
+#define FLOWER_ATTR_DST_MAC_MASK    1 << 0x7
+#define FLOWER_ATTR_SRC_MAC         1 << 0x8
+#define FLOWER_ATTR_SRC_MAC_MASK    1 << 0x9
+#define FLOWER_ATTR_IP_DSCP         1 << 0xa
+#define FLOWER_ATTR_IP_DSCP_MASK    1 << 0xb
+#define FLOWER_ATTR_PROTO           1 << 0xc
+/** @endcond */
+
+#define FLOWER_DSCP_MAX             0xe0
+#define FLOWER_DSCP_MASK_MAX        0xe0
+#define FLOWER_VID_MAX              4095
+#define FLOWER_VLAN_PRIO_MAX        7
+
+static struct nla_policy flower_policy[TCA_FLOWER_MAX + 1] = {
+        [TCA_FLOWER_KEY_ETH_TYPE]	= { .type = NLA_U16 },
+        [TCA_FLOWER_KEY_ETH_DST]	= { .maxlen = ETH_ALEN },
+	[TCA_FLOWER_KEY_ETH_DST_MASK]	= { .maxlen = ETH_ALEN },
+	[TCA_FLOWER_KEY_ETH_SRC]	= { .maxlen = ETH_ALEN },
+	[TCA_FLOWER_KEY_ETH_SRC_MASK]	= { .maxlen = ETH_ALEN },
```

Comment:
the MAC addresses are all `ETH_ALEN` bytes. Please also set `.minlen = ETH_ALEN`.